alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+use std::cmp;
+
+use glutin::Rect;
+
+use alacritty_terminal::selection::SelectionRange;
+use alacritty_terminal::term::{DamagedLine, RenderableCursor, SizeInfo, TermDamage};
+use alacritty_terminal::vi_mode::ViModeCursor;
+
+/// Minimum part of damaged square when intersecting damaged rectangles.
+pub const MIN_DAMAGED_SQUARE_PART: f32 = 0.4;
+
+/// An error when we for look ahead in iterator, used to try to collect rectangles which pretty
+/// close to each other, but don't fit into `MIN_DAMAGED_SQUARE_PART` just by looking on the first
+/// next damaged rectangle.
+pub const MIN_DAMAGED_SQUARE_PART_ERR: f32 = MIN_DAMAGED_SQUARE_PART / 2.;
+
+pub struct RendererDamage<'a> {
+    index: usize,
+    rect: Option<Rect>,
+    size_info: SizeInfo<u32>,
+    terminal_damage: &'a mut Vec<DamagedLine>,
+}
+
+impl<'a> RendererDamage<'a> {
+    pub fn new(
+        terminal_damage: &'a mut TermDamage,
+        size_info: SizeInfo<u32>,
+        cursor: RenderableCursor,
+        vi_cursor: Option<ViModeCursor>,
+        selection: Option<SelectionRange>,
+        display_offset: usize,
+    ) -> Self {
+        let fully_damaged = terminal_damage.take_fully_damaged();
```

Comment:
I really don't like the idea of using a boolean check method to also mutate state. That just doesn't sit right with me.

Should we maybe have a `damage` method on the terminal that returns an enum containing either `Full`, or `Partial(Vec<DamagedLine>)`? Seems like a much more proper solution. Then we can just manually clear damage.

Maybe make that method clear all damage too, then it would at least be a reasonable option. But I'm not sure I would.

Reply:
I think I've got an idea how to rewrite it a bit, so it'll make more sense.